nbgrader/preprocessors/instantiatetests.py

```diff
@@ -0,0 +1,420 @@
+    ).tag(config=True)
+
+    enforce_metadata = Bool(
+        True,
+        help=dedent(
+            """
+            Whether or not to complain if cells containing autotest delimiters
+            are not marked as grade cells. WARNING: disabling this will potentially cause
+            things to break if you are using the full nbgrader pipeline. ONLY
+            disable this option if you are only ever planning to use nbgrader
+            assign.
+            """
+        )
+    ).tag(config=True)
+
+    comment_strs = Dict(
+        key_trait=Unicode(),
+        value_trait=Unicode(),
+        default_value={
+            'ir': '#',
+            'python': '#',
+            'python3': '#'
+        },
+        help=dedent(
+            """
+            A dictionary mapping each Jupyter kernel's name to the comment string for that kernel.
+            For an example, one of the entries in this dictionary is "python" : "#", because # is the comment
+            character in python.
+            """
+        )
+    ).tag(config=True)
+
+    sanitizers = Dict(
+        key_trait=Unicode(),
+        value_trait=Callable(),
+        default_value={
+            'ir': lambda s: re.sub(r'\[\d+\]\s+', '', s).strip('"').strip("'"),
+            'python': lambda s: s.strip('"').strip("'"),
+            'python3': lambda s: s.strip('"').strip("'")
+        },
+        help=dedent(
+            """
+            A dictionary mapping each Jupyter kernel's name to the function that is used to
+            sanitize the output from the kernel within InstantiateTests.
+            """
+        )
+    ).tag(config=True)
+
+    sanitizer = None
+    kernel_name = None
+    kc = None
+    execute_result = None
+
+    def preprocess(self, nb, resources):
+        # avoid starting the kernel at all/processing the notebook if there are no autotest delimiters
+        for index, cell in enumerate(nb.cells):
+            # look for an autotest delimiter in this cell's source; if we find one, process this notebook
+            # short-circuit ignore non-code cells
+            if (cell.cell_type == 'code') and (self.autotest_delimiter in cell.source):
+                # get the kernel name from the notebook
+                kernel_name = nb.metadata.get("kernelspec", {}).get("name", "")
+                if kernel_name not in self.comment_strs:
+                    raise ValueError(f"Kernel {kernel_name} has not been specified in InstantiateTests.comment_strs")
+                if kernel_name not in self.sanitizers:
+                    raise ValueError(f"Kernel {kernel_name} has not been specified in InstantiateTests.sanitizers")
+                self.log.debug(f"Found kernel {kernel_name}")
+                resources["kernel_name"] = kernel_name
+
+                # load the template tests file
+                self.log.debug('Loading template tests file')
+                self._load_test_template_file(resources)
```

Comment:
Since the autotests is optional, is it possible to avoid running the preprocessor if the config file does not exist ?

Reply:
Ah yes -- this is related to a new change I made compared with the old PR. I call `_load_test_template_file` in `preprocess` now, but only after checking if `self.autotest_delimiter in cell.source`. So indeed we can catch an error sooner.

(But we should raise an exception here. If there is a `self.autotest_delimiter` in a cell, it's clear the instructor wants autotest to run -- moving on silently without running autotest might create notebooks without tests the instructor expects to be there.)

Reply:
I missed the check about `self.autotest_delimiter in cell.source`. Anyway, it's probably better to raise if the file doesn't exist, rather than warning of something that will happen anyway (since *AUTOTEST* has already been found in the file).